### PR TITLE
Exercise sheet for Spack exercise

### PR DESCRIPTION
**Pull request by ajaust**
This PR contains the CPack and Spack exercise which is currently also in #6 

The file will be removed from #6 as soon as the comments there have been clarified.

### CODE REVIEW

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+Please use a suitable Docker container for this task. You should be able to reuse the Docker container from last week's exercise. The Docker container will be especially important if you want to work on the optional task
+
+### Workflow (CPack)
+
+- Create a fork of the [GitHub repository](https://github.com/Simulation-Software-Engineering/cpack-exercise).
+- Adapt and create the files necessary to create packages via CPack.
+- Also add your `Dockerfile` to the repository.
+- Submit your code via a pull request. You can also open a (draft) pull request before you are incorporated all changes. In this case, do not forget to mark the pull request as ready to review when you are finished.
+    - The pull request should be named `[USERNAME] CPack exercise`, e.g., `[jaustar] CPack exercise`.
```
Comment (ajaust) [inline building-and-packaging/material/packaging_debian_and_spack_exercise.md:31]: @uekerman suggestion was the following:
```suggestion
    - The pull request should be named `[USERNAME] CPack exercise`, e.g., `[ajaust] CPack exercise`.
```

My reply was:
> I normally prefer GitLab usernames, but if we used the GitHub username for submissions so far I am fine with that as well.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+Additionally, for the Debian package, make sure that
+
+- the executable is stripped, i.e., debug symbols should removed from the executable.
+- the list of dependencies is generated automatically for the Debian package.
+- the Debian package is generated with default file naming scheme.
+
+Please use a suitable Docker container for this task. You should be able to reuse the Docker container from last week's exercise. The Docker container will be especially important if you want to work on the optional task
+
+### Workflow (CPack)
```
Comment (ajaust) [inline building-and-packaging/material/packaging_debian_and_spack_exercise.md:25]: @uekerman 's comment was

> Is "workflow" the correct word here?

My reply was

> You did (not) complain about it yet in the Spack section. smile
> 
> What would you prefer/suggest?

**Review by ajaust — COMMENTED** [review]
I moved over review comments for the exercise sheet from #6.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md (hunk 2 above)
Comment (uekerman) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md:25]: Will complain in the spack section as well :wink: 

Mmh, maybe "submission steps" or "steps to submission" ?

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+- Create a Spack package for the Python code that you have packaged with pip. Checkout [Spack's documentation regaring the `PythonPackage`](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html)  class for recipes if you want to work on this.
+
+## Remarks and Tips
+
+- If you need to specify a homepage, you can use the course homepage `https://simulation-software-engineering.github.io/homepage/`.
+- All parts of the exercise should be ideally done on Docker container based on Ubuntu 20.04.
+- If you are in a container and your libraries in `/usr/local/lib` are not found, you can extend the `LD_LIBRARY_PATH` accordingly.
+- If you need an inspiration on how to write more complicated Spack recipes, you can check packages that are already shipped with spack (`spack edit PACKAGENAME` or browse).
+- If you have `Boost` or `deal.ii` installed in the Docker container via `apt`, finding pre-installed packages using `spack external find` might fail. In this case you have to define the dependencies manually or run `spack external find` before installing the critical packages.
```
Comment (uekerman) [inline building-and-packaging/material/packaging_debian_and_spack_exercise.md:96]: Let's remove the deal.ii remark here. Might only confuse.

**Review by uekerman — COMMENTED** [review]
Some feedback on the spack exercise. I have not yet done all steps, but seems rather clear to me what to do. I will try the final version next week again.

* What is the WALBERLA comment in `01_nodoc` doing? :smile: 
* Because of this file, the Docker container currently needs to be build from the `docker` folder.
* Some comments in `Dockerfile` seem to be at the wrong locations
* Currently, there are two options which Docker image to use: To build with the Dockerfile you provide or to download from Dockerhub. These two options are quite mixed up in the description. Why the second option "Dockerhub"? If you really want to keep both, we should explain better that there is a choice to make.

**Comment by ajaust** [comment]
This PR should only be used for the Spack exercise review. The CPack exercise will go into another PR.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md (hunk 2 above)
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: Fixed

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+Now we want to package [another code](https://github.com/Simulation-Software-Engineering/spack-exercise) from Github. It is an adapted version of the previous week's exercise again. However, the `deal.ii` dependency has been removed as compiling `deal.ii` from scratch takes a very long time. Additionally, several releases with different depencies have been added.
+
+The default name of your Spack package will be `spack-exercise` while the executable it produces is called `spackexample`. It also creates a library  `libspackexamplelib` (what a name) which is needed to run `spackexample`. Under normal circumstances you do not need to edit the CMake configuration for this exercise.
+
+The [code repository](https://github.com/Simulation-Software-Engineering/spack-exercise) contains three releases:
+
+- `v0.1.0` has no special dependencies. The C++ standard library is automatically included.
+- `v0.2.0` depends on Boost (at least version `1.65.1`)
+- `v0.3.0` depends on Boost (at least version `1.65.1`) and `yaml-cpp` (at least version `0.70.0`)
```
[the remark this answers is not in the document]
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: Fixed

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+Create a Spack package for all releases the respect the dependencies appropriately. Make sure you also add yourself as maintainer to your package. Your final package should not contain any `FIXME` parts anymore.
+
+### Workflow (Spack)
+
+- Create a fork of the [GitHub repository](https://github.com/Simulation-Software-Engineering/spack-exercise).
+- Adapt and create files as necessary to create the package via Spack.
+- Please add a directory name `spack-exercise` inside the `spack` directory of the repository.
```
[the remark this answers is not in the document]
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: I will rename/restructure this. Ideally this directory would contain the directory `spack-exercise` which in turn would contain a `package.py` file.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md (hunk 3 above)
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: Fixed

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    - Inside the `spack-exercise` repository you should add the `package.py` of your Spack package.
+- Please also add the `Dockerfile` you have used for the exercise if it has been adapted from the initial `Dockerfile` (see below). This `Dockerfile` should be in the root of the repository. Also add all other files you needed to create/edit to solve this task.
+- Submit your code via a pull request. You can also open a (draft) pull request before you are incorporated all changes. In this case, do not forget to mark the pull request as ready to review when you are finished.
+    - The pull request should be named `[USERNAME] Spack exercise`, e.g., `[jaustar] Spack exercise`.
+
+It is strongly recommended to use work inside a Docker container. The directory `docker` includes a `Dockerfile` one may use as starting point for the exercise. It will come with Spack preinstalled (but **not** yet completely set up yet). Inside this container you are the `spackbuilder` user who does not have superuser rights. If you need superuser rights inside the container you can use the `sudo` command.
```
[the remark this answers is not in the document]
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: Fixed.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md (hunk 1 above)
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: This is specified explicitly now in the exercise sheet.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+- Also add your `Dockerfile` to the repository.
+- Submit your code via a pull request. You can also open a (draft) pull request before you are incorporated all changes. In this case, do not forget to mark the pull request as ready to review when you are finished.
+    - The pull request should be named `[USERNAME] CPack exercise`, e.g., `[jaustar] CPack exercise`.
+
+
+## 2. Spack
+
+**Attention/TODO**: Use [test repository](https://github.com/Simulation-Software-Engineering/spack-exercise) for pushing solutions. You will still have to build from the [public repository](https://github.com/Simulation-Software-Engineering/spack-exercise).
+
+Now we want to package [another code](https://github.com/Simulation-Software-Engineering/spack-exercise) from Github. It is an adapted version of the previous week's exercise again. However, the `deal.ii` dependency has been removed as compiling `deal.ii` from scratch takes a very long time. Additionally, several releases with different depencies have been added.
```
[the remark this answers is not in the document]
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: This has been rephrased.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+The default name of your Spack package will be `spack-exercise` while the executable it produces is called `spackexample`. It also creates a library  `libspackexamplelib` (what a name) which is needed to run `spackexample`. Under normal circumstances you do not need to edit the CMake configuration for this exercise.
+
+The [code repository](https://github.com/Simulation-Software-Engineering/spack-exercise) contains three releases:
+
+- `v0.1.0` has no special dependencies. The C++ standard library is automatically included.
+- `v0.2.0` depends on Boost (at least version `1.65.1`)
+- `v0.3.0` depends on Boost (at least version `1.65.1`) and `yaml-cpp` (at least version `0.70.0`)
+
+Create a Spack package for all releases the respect the dependencies appropriately. Make sure you also add yourself as maintainer to your package. Your final package should not contain any `FIXME` parts anymore.
```
[the remark this answers is not in the document]
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: Done.

File: building-and-packaging/material/packaging_debian_and_spack_exercise.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajaust) [reply building-and-packaging/material/packaging_debian_and_spack_exercise.md]: This part is gone now. I just want the file `package.py`. I hope that is simpler to do and to understand.

**Comment by ajaust** [comment]
I think I have included all your remarks now. I was messing around to much so I could not take your suggestions via GitHub anymore, but I made sure thare is at least one commit with you as co-author so you can also get the GitHub credit.

> Some feedback on the spack exercise. I have not yet done all steps, but seems rather clear to me what to do. I will try the final version next week again.
> 
>     * What is the WALBERLA comment in `01_nodoc` doing? smile

I assume that is the repository where this file comes from. I removed it for now as there was no copyright statement in the WALBERLA repository. I think that WALBERLA is the Lattice-Boltzmann solver of the Rüde group in Erlangen.

>     * Because of this file, the Docker container currently needs to be build from the `docker` folder.

I specify now that the students should use the prebuild image [`ajaust/spack-package-exercise`](https://hub.docker.com/repository/docker/ajaust/spack-package-exercise). This image is still quite large (almost 1 GB), but it comes with a preconfigured Spack including Boost. So there should be no/less time wasted with compiling things and setting up Spack.

>     * Some comments in `Dockerfile` seem to be at the wrong locations

I fixed/changed/removed. the comments.

>     * Currently, there are two options which Docker image to use: To build with the Dockerfile you provide or to download from Dockerhub. These two options are quite mixed up in the description. Why the second option "Dockerhub"? If you really want to keep both, we should explain better that there is a choice to make.

As mentione above, I request them to use my image [`ajaust/spack-package-exercise`](https://hub.docker.com/repository/docker/ajaust/spack-package-exercise) now.

**Pull request merged**